### PR TITLE
Fixing #5500 (missing test in return clause of match leading to anomaly)

### LINKED ISSUE #5500: Invalid_argument("splay_arity")
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#5500
From: @thierry-martinez
Reported version: 8.6
CC:   @herbelin

See also: [BZ#5547](https://github.com/coq/coq/issues?q=is%3Aissue%20%22Original%20bug%20ID%3A%20BZ%235547%22)

### PR DESCRIPTION
**Pull request by herbelin**
The anomaly should at least be an error, and this is what this PR does (see commit message for details). 

Otherwise, with generalized small inversion we shall eventually be able to produce a solution to the pattern-matching problem reported in the bug report which does not fail (see the report [#5500](https://coq.inria.fr/bugs/show_bug.cgi?id=5500) for details).

Fixes #5500.

### CODE REVIEW

**Comment by maximedenes** [comment]
Looks good to me. Rerunning tests that failed for unrelated reasons, and should be good to merge.

**Comment by maximedenes** [comment]
This seems to break fiat parsers, although I'm not completely sure it is related.

**Comment by herbelin** [comment]
@maximedenes: I don't see from what it could come else (travis is again trustable at the current time, right?). I'll investigate.

**Comment by maximedenes** [comment]
Travis builds trunk and 8.6 fine at the moment, but we didn't close and reopen the PR, so it is not impossible that an earlier problem remained here. But it is unlikely, I don't remember any recent fiat-parsers issues.

**Comment by SkySkimmer** [comment]
- travis just the first commit https://travis-ci.org/SkySkimmer/coq/jobs/236932231 works
- travis both commits https://travis-ci.org/SkySkimmer/coq/jobs/236931520 fails
- travis merge into trunk https://travis-ci.org/SkySkimmer/coq/jobs/236933088 fails

**Comment by herbelin** [comment]
I started to investigate it but did not finish. It fails on some `_subproof` generated by `abstract` not found in the typing environment. So, suspectingly an `abstract`-related bug that the extra call to the type-checker brought by the PR makes visible. I'm unsure that the `abstract` bug can be fixed quickly. Maybe woult it be easier (and safer) to change `fiat`, at least temporarily, rather than piling bugs.

**Comment by Zimmi48** [comment]
@herbelin Since this is a bugfix, this is safe to postpone after the freeze (but preferably before the beta) if you have other more urgent matters.

**Comment by herbelin** [comment]
@JasonGross: The fix for [BZ#5547](https://coq.inria.fr/bugs/show_bug.cgi?id=5547) is breaking fiat-parsers in that it reveals a bad interaction of `abstract` and type classes as reported in [BZ#5703](https://coq.inria.fr/bugs/show_bug.cgi?id=5703). Fixing this latter bug would require heavy changes, so not before 8.8. Would you prefer to have BZ#5547 integrated now or change fiat-parsers to circumvent the BZ#5703 bug. This fails at line 15 of file [SharpenedABStarParseTree.v](https://github.com/JasonGross/fiat/blob/master/src/Parsers/Refinement/SharpenedABStarParseTree.v#L15), but I did not explore how to avoid the bug (maybe removing the `_opaque` suffix?).

**Comment by herbelin** [comment]
You don't use explicitly `type of` but the fix to BZ#5547 adds an internal "type of" when building a `match`. The failing `match` is the one constructed at the end of `make_parser_informative_opaque` in `ReductionTactics.v`. In the return clause of the match, there is a variable `b` which is made by `make_Parser`, which forces type classes to builds an abstract proof with the `Hint Extern` trick on `eq_refl_vm_cast`. A fix could be to remove the abstract in the `eq_refl_vm_cast` for `make_Parser`, but maybe it is asking too much.

As far as I'm concerned, I wouldn't say that it is "needed" to integrate the BZ#5587 fix to 8.7, even more that 8.8 shall come quickly.

**Comment by JasonGross** [comment]
@herbelin is replacing `b` with `b : bool` and/or replacing `b'` with `b' : bool` everywhere (or somewhere) in that `match` enough to fix the issue, as suggested by the discussion on bugzilla?  If so, could you make such a PR against fiat, and I'll merge it?

**Comment by maximedenes** [comment]
@herbelin please update the milestone to reflect your current schedule on this. Thanks!

**Comment by maximedenes** [comment]
@herbelin ping

**Comment by herbelin** [comment]
@maximedenes: this should be merged. The failing example in fiat (which is not due to this PR but which reveals another bug) is bit. If @jasongross proposal works, that may be quick to do the fiat overlay. If not, that would require more time (for the overlay, not for the PR itself).

More generally, I plan to do all needed relevant rebases of my PR today.

**Comment by maximedenes** [comment]
Ok, thanks for the update!

**Comment by maximedenes** [comment]
Failed at https://travis-ci.org/herbelin/github-coq/jobs/350961804 even with the overlay.

So postponing to 8.8.0 (it is a bug fix, so it can be integrated later).

**Comment by herbelin** [comment]
Where did you get the failure? In the log you're showing, the overlay has not been applied.

Note that even if a bugfix, we should have worked at integrating it much more earlier. So, let's not postpone it once more for bad reasons.

**Comment by maximedenes** [comment]
> In the log you're showing, the overlay has not been applied.

Well then, why? It is testing commit https://github.com/herbelin/github-coq/commit/ef6000bcb622160f7b95cc93853bc2c605e27c57 which is the last this PR contains AFAICT.

**Comment by herbelin** [comment]
OK, my bad, I forgot the `.sh` suffix to the overlay. I pushed a fix.

**Comment by maximedenes** [comment]
Travis fails with some warnings, postponing to 8.8.0 since it is a bugfix.

**Comment by Zimmi48** [comment]
So it's fine to merge bugfixes with overlays after the release of the beta?

**Comment by Zimmi48** [comment]
The deprecation warning should be easily fixable BTW.

**Comment by maximedenes** [comment]
> with overlays

hum...

**Comment by maximedenes** [comment]
@herbelin can you fix the warnings ASAP so that this gets a last chance?

**Comment by herbelin** [comment]
I fixed the `family_of_sort` warning.

**Comment by maximedenes** [comment]
Pipeline running here: https://gitlab.com/maximedenes/coq/pipelines/18663911

**Comment by maximedenes** [comment]
fiat parser failed again: https://travis-ci.org/herbelin/github-coq/builds/351391833

bailing out for this one

**Comment by maximedenes** [comment]
Maybe the fiat parser fix can be made backward compatible?

**Comment by herbelin** [comment]
Well, testing again (but I'm close to capitulate...).

**Comment by maximedenes** [comment]
I don't think we need to rush this. Especially if the fix can be made backward compatible, I don't see a problem in integrating this after the beta.

**Comment by herbelin** [comment]
OK, got it. I made another typo (missing an `s` to `parser` in the overlay).

**Comment by Zimmi48** [comment]
@herbelin Can you rebase and push? This should ensure that the currently failing CI, this should also fix the whitespaces (if you have the pre-commit hook, or maybe you need to use the `--whitespace=fix` option), and this should request reviews from code owners.

**Comment by maximedenes** [comment]
@mattam82 do you plan to have a look at this or should I update the milestone?

**Comment by maximedenes** [comment]
@mattam82 ping

**Comment by maximedenes** [comment]
@mattam82 I got no news, so delaying to 8.8.1. If you plan to review it today, please let me know ASAP. Thanks!

**Comment by herbelin** [comment]
@mattam82: Idem, if you are missing factual time, maybe you can propose that someone else take care of it.

**Comment by Zimmi48** [comment]
Very sad that this missed 8.8.0. Antoine just encountered this anomaly again. Unfortunately since this fix is a breaking change the delaying to 8.8.1 was wrong and we have to delay instead to 8.9+beta1.

**Review by mattam82 — APPROVED** [review]
Ok, the code looks fine to me. IIUC the failure, e.g. the for the `refine` at the end of 5547.v is because the check we have is too approximate and conservative?
EDIT: ~~Otherwise, is the fiat-parsers problem now solved by the overlay?~~ Apparently yes.

**Comment by mattam82** [comment]
Ok, I think I understand that 5547.v fails because the nested pattern-matching algorithm itself creates an ill-typed term and we should rather error here.

**Comment by mattam82** [comment]
@herbelin can you submit https://github.com/herbelin/fiat/commit/dce379fc7a37cbc112f79610bb8ea033fbc34506 upstream please?

**Comment by Zimmi48** [comment]
IIUC the overlay had never been submitted to fiat, let alone merged, so I opened a PR for this https://github.com/mit-plv/fiat/pull/12.
In general we prefer to have backward compatible patches merged *before* we merge the PR on Coq.

**Comment by herbelin** [comment]
Ah, sorry, it was a backward-compatible overlay but I forgot to submit it. Thanks a lot @Zimmi48.

**Comment by mattam82** [comment]
I had a doubt about backward compatibility actually, thanks for clearing it up.

**Comment by SkySkimmer** [comment]
Did we ever bench this?

**Comment by mattam82** [comment]
I didn’t see a bench at least.
Le jeu. 14 juin 2018 à 13:16, Gaëtan Gilbert <notifications@github.com> a
écrit :

**Comment by herbelin** [comment]
Basically, there is an extra `type_of` in compilation of `match`. How could it be noticeable in a sea of time spent in tactics, or even among the rest of typing or unification problems needed to type the term containing the `match`?

**Pull request merged**
